gobblin-data-management/src/main/java/gobblin/data/management/retention/action/AccessControlAction.java

```diff
@@ -0,0 +1,133 @@
+    this.owner = Optional.fromNullable(ConfigUtils.getString(actionConfig, OWNER_KEY, null));
+    this.group = Optional.fromNullable(ConfigUtils.getString(actionConfig, GROUP_KEY, null));
+    this.selectionPolicy = createSelectionPolicy(actionConfig);
+  }
+
+  /**
+   * Applies {@link #selectionPolicy} on <code>allVersions</code> and modifies permission/owner to the selected {@link DatasetVersion}s
+   * where necessary.
+   * <p>
+   * This action only available for {@link FileSystemDatasetVersion}. It simply skips the operation if a different type
+   * of {@link DatasetVersion} is passed.
+   * </p>
+   * {@inheritDoc}
+   * @see gobblin.data.management.retention.action.RetentionAction#execute(java.util.List)
+   */
+  @Override
+  public void execute(List<DatasetVersion> allVersions) throws IOException {
+
+    // Select version on which access control actions need to performed
+    for (DatasetVersion datasetVersion : this.selectionPolicy.listSelectedVersions(allVersions)) {
+
+      // Perform action if it is a FileSystemDatasetVersion
+      if (datasetVersion instanceof FileSystemDatasetVersion) {
+        FileSystemDatasetVersion fsDatasetVersion = (FileSystemDatasetVersion) datasetVersion;
+        for (Path path : fsDatasetVersion.getPaths()) {
+
+          if (this.fs.exists(path)) {
+
+            // Update permissions if set in config
+            if (this.permission.isPresent()) {
+              FsPermission fsPermission = new FsPermission(this.permission.get());
+              this.fs.setPermission(path, fsPermission);
+              log.info(String.format("Set permissions for %s to %s", path.toString(), fsPermission));
+            }
+
+            // Update owner and group if set in config
+            if (this.owner.isPresent() || this.group.isPresent()) {
+              this.fs.setOwner(path, this.owner.orNull(), this.group.orNull());
+              log.info(String.format("Set owner and group for %s to %s:%s", path.toString(), this.owner.orNull(),
+                  this.group.orNull()));
+            }
+          }
+        }
+      }
+    }
+  }
+
+  /*
+   * Since {@link VersionSelectionPolicy} does not have a factory to create new objects we need to use the legacy
+   * pattern of creating new objects using GobblinConstructorUtils
+   */
+  @SuppressWarnings("unchecked")
+  private static VersionSelectionPolicy<DatasetVersion> createSelectionPolicy(Config selectionConfig) {
+    try {
+      String selectionPolicyKey =
+          StringUtils.substringAfter(ConfigurableCleanableDataset.SELECTION_POLICY_CLASS_KEY,
+              ConfigurableCleanableDataset.CONFIGURATION_KEY_PREFIX);
+      Preconditions.checkArgument(selectionConfig.hasPath(selectionPolicyKey));
+      String className = selectionConfig.getString(selectionPolicyKey);
+      return (VersionSelectionPolicy<DatasetVersion>) GobblinConstructorUtils.invokeFirstConstructor(
+          Class.forName(className), ImmutableList.<Object> of(selectionConfig),
```

Comment:
Yes the `ClassAliasResolver` does not cache aliasToClass mapping. I will have setup a cache so that we don't have to scan the classpath for every `Dataset`.